### PR TITLE
Replace bandit, flake8, isort, pyupgrade, and yesqa with ruff

### LINKED ISSUE #1227: Code health tracker
## 🚀 Feature

List including the pending things to do to improve the package health

- [x] Add codespell in pre-commit
- [x] Add pyupgrade in pre-commit
- [x] Add dependabot -> already enabled!
- [x] Deepsource: https://deepsource.io/gh/kornia/kornia (need to find a proper way to automate)
- [x] docformatter: https://github.com/myint/docformatter
- [x] yesqa: https://github.com/asottile/yesqa
- [ ] full support to `bandit` -> https://github.com/kornia/kornia/issues/1245
- [x] configure dependabot to `requirements-dev.txt`
- [ ] flake8-bugbear https://github.com/kornia/kornia/pull/1233
- [ ] Pytest raises 600+ warnings https://github.com/kornia/kornia/issues/1225
- [ ] Tests take a long time https://github.com/kornia/kornia/discussions/1260

### PR DESCRIPTION
**Pull request by cclauss**
#### Changes

Fixes #1227

[Ruff](https://beta.ruff.rs/) supports [over 500 lint rules](https://beta.ruff.rs/docs/rules) and can be used to replace [Flake8](https://pypi.org/project/flake8/) (plus dozens of plugins), [isort](https://pypi.org/project/isort/), [pydocstyle](https://pypi.org/project/pydocstyle/), [yesqa](https://github.com/asottile/yesqa), [eradicate](https://pypi.org/project/eradicate/), [pyupgrade](https://pypi.org/project/pyupgrade/), and [autoflake](https://pypi.org/project/autoflake/), all while executing (in Rust) tens or hundreds of times faster than any individual tool.

The `ruff` Action uses minimal steps to run in ~5 seconds, rapidly providing intuitive GitHub Annotations to contributors.

image

1. I left `pydocstyle` in place because I could not master https://beta.ruff.rs/docs/settings/#pydocstyle
2. I do not yet know `.devcontainer` so I made no changes there (yet!)
3. I did not add the GitHub Action to have `ruff` generate GitHub Annotations but that is easy and fast to add.

#### Type of change

- [ ] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/nerf/camera_utils.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ def __init__(self, line: str) -> None:
 
         # Read line with camera quaternion
         line = line.strip()
-        if line.endswith('png') or line.endswith('jpg\n'):
+        if line.endswith(('png', 'jpg\n')):
```
[the remark this answers is not in the document]
Comment (cclauss) [reply kornia/nerf/camera_utils.py:70]: Actually `line` will ___never___ end with a \n after it has been stripped!!

**Comment by edgarriba** [comment]
@johnnv1 feel free to merge

**Comment by johnnv1** [comment]
> @johnnv1 feel free to merge

we can go with Ruff right after we do the next release 😄

File: requirements/requirements-dev.txt [hunk 2]
```diff
@@ -7,4 +6,5 @@ pre-commit>=2
 pydocstyle
 pytest==7.3.1
 pytest-cov==4
+ruff==0.0.259
```
Comment (johnnv1) [inline requirements/requirements-dev.txt:9]: pre-commit will take care ruff versioning
```suggestion
```

File: pyproject.toml [hunk 3]
```diff
@@ -129,3 +120,51 @@ D204
...
+
+[tool.ruff.pylint]
+max-args = 16
+max-branches = 45
+max-returns = 13
+max-statements = 258
+
+[tool.ruff.per-file-ignores]
+"*/__init__.py" = ["F401"]
+"docs/*" = ["S101", "S113"]
```
Comment (johnnv1) [inline pyproject.toml:162]: this fixes the merge issue with the docs examples
```suggestion
"docs/*" = ["S101", "S113"]
"docs/generate_example_images.py" = ["C901", "PLR0912", "PLR0915"]
```

File: pyproject.toml [hunk 4]
```diff
@@ -129,3 +120,51 @@ D204
...
+"examples/train/object_detection/main.py" = ["E741"]
+"kornia/__init__.py" = ["I001"]
+"kornia/augmentation/container/base.py" = ["PLC1901"]
+"kornia/color/hls.py" = ["E741"]
```
Comment (johnnv1) [inline pyproject.toml:166]: this fixes the merge issue with SAM
```suggestion
"kornia/color/hls.py" = ["E741"]
"kornia/contrib/models/sam/architecture/image_encoder.py" = ["PLR0913"]
```

**Review by johnnv1 — APPROVED** [review]

**Pull request merged**